Adds the ability for users to invoke the soft keyboard anywhere, at any time, even on screens that do not normally allow keyboards. This approach works even on Android 13 (API 33)

LostRuins opened the pull request:
I went ahead and did it myself, because I really needed the functionality. I am quite certain no other FOSS keyboard has this ability working in Android 13.

This PR adds the ability for users to invoke the soft keyboard anywhere, at any time, even on screens that do not normally allow keyboards. This approach works even on Android 13 (API 33), and is an essential replacement for the now-defunct Hacker Keyboard's Permanent Notification, which had a similar capability but stopped functioning after API 30 (https://github.com/klausw/hackerskeyboard/issues/897).

Closes https://github.com/Julow/Unexpected-Keyboard/issues/496, Closes https://github.com/Julow/Unexpected-Keyboard/issues/71, Closes https://github.com/Julow/Unexpected-Keyboard/issues/392 and Closes https://github.com/Julow/Unexpected-Keyboard/issues/69 

How to use:
1. Launch Unexpected Keyboard, and click `Show Persistent Keyboard`
2. Grant the `Draw Over Other Apps` permission, allowing overlays to be displayed.
3. Now, whenever you require a keyboard, launch the app and press `Show Persistent Keyboard`
4. A button overlay will appear in the top left corner
5. Navigate back to the app you want to use that does not support keyboards.
6. Click the `Show` button in the top left corner. The overlay will be closed, and at the same time the soft keyboard will be invoked.
7. Enjoy your ability to use a keyboard anywhere, like Richard Stallman intended.

![demo](https://github.com/Julow/Unexpected-Keyboard/assets/39025047/8c1df984-98b0-430e-974d-f5f64b00c10b)

LostRuins commented:
Feel free to modify or change anything, I am not an android dev so I might not be following best practices.

Julow reviewed (commented):
This is fantastic, thanks! I'll make a proper review and merge this soon.

Only APKs built by the CI are trustworthy, here's the latest for this PR: https://github.com/Julow/Unexpected-Keyboard/actions/runs/7049282572?pr=497

LostRuins commented:
Worth noting that on android 13, it seems that there must be at least one navigable ui element for the keyboard to remain displayed. Otherwise, it gets dismissed. If you would have been able to plug a real keyboard and navigate that app via arrow keys, then it would probably work.

Spike-from-NH commented:
I'm glad you took the initiative!
① Can I drag this button somewhere else if it occludes something important?
② The "X" box is to dismiss the button? Can you jam them together to better show they are parts of the same control? If dismissed, can they be re-enabled in the way they were originally?
③ If Draw Over Other Apps is not granted, do you ask for it?
④ Add the enable button to the Settings menu as well?
⑤ Does it look good in all Themes?

Julow commented on res/layout/overlay_button.xml:
```diff
@@ -0,0 +1,19 @@
+        android:layout_width="32dp"
+        android:layout_height="32dp"
+        android:background="#FF8282"
+        android:text="x" />
```
The close button takes a lot of place and might not be useful. The show button could turn into the close button while the keyboard is showing.

Julow commented on res/layout/launcher_activity.xml:
```diff
@@ -6,4 +6,5 @@
   <TextView style="@style/paragraph" android:text="https://github.com/Julow/Unexpected-Keyboard" android:autoLink="web" android:linksClickable="true"/>
   <TextView android:id="@+id/launcher_tryhere_text" style="@style/paragraph" android:text="@string/launcher_tryhere"/>
   <EditText android:id="@+id/launcher_tryhere_area" style="@style/paragraph" android:inputType="text"/>
+  <Button style="@style/paragraph" android:text="Show Persistent Keyboard" android:onClick="startOverlayService" android:layout_width="wrap_content"/>
```
Opening the app and clicking this button is a lot of work. Any hope of having a permanent notification instead ?

Julow commented on res/layout/overlay_button.xml:
```diff
@@ -0,0 +1,19 @@
+        android:id="@+id/invokeKeyboardButton"
+        android:layout_width="72dp"
+        android:layout_height="wrap_content"
+        android:text="Show" />
```
This buttons shows up on top of every other apps. "Show" doesn't mean much without context. The app's icon or a keyboard symbol would be more intuitive.

Julow commented on srcs/juloo.keyboard2/LauncherActivity.java:
```diff
@@ -26,6 +29,39 @@ public void onCreate(Bundle savedInstanceState)
+  private static final int REQUEST_CODE_OVERLAY_PERMISSION = 1601;
+  private void requestOverlayPermission() {
+    // Request the permission using an Intent
+    Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getPackageName()));
```
`Uri.parse` can be replaced by `Uri.fromParts("package", getPackageName(), "")`.

Julow commented on srcs/juloo.keyboard2/LauncherActivity.java:
```diff
@@ -26,6 +29,39 @@ public void onCreate(Bundle savedInstanceState)
+    if (requestCode == REQUEST_CODE_OVERLAY_PERMISSION) {
+      // Check if the permission was granted
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && Settings.canDrawOverlays(this)) {
+        Toast.makeText(this, "Overlay permissions are granted. You can now launch persistent keyboards.", Toast.LENGTH_SHORT).show();
```
We could start the OverlayService here.

Julow reviewed (commented):
This doesn't work on my device, the keyboard opens then closes immediately on top of any app. Any idea on how to debug this ?

LostRuins commented:
Hi @Julow, I have a bit of bad news. It seems like I was mistaken on the specifics this approach.

If the *user's device* is below Android 13, it will work in _all_ cases.
If the *user's device* is at or above Android 13, and the *target SDK* of the app is below API 33, it will _also_ work.
However, if the APK target SDK exceeds 33, **and** the device is **also** android 13 or above, then this approach does not work anymore, as the keyboard will be closed the moment the view loses focus.

You can try by either using any older device, or downloading some random older app from fdroid. I did not notice this at first, as many of the apps installed on my device are older APKs. 

With this disappointing news, I am not sure if this warrants closing this PR or not, it's your choice... perhaps it could still be of some use either for older apps or older devices like in my own use case.

Nevertheless, I have tidied up the PR a bit, so do take a look. 
- The button has been changed to an ImageView with the app logo as an icon.
- The dismiss button has been resized and changed to be more intuitive looking.
- They overlay is now persistent and does not disappear on a single use. Instead, the keyboard is invoked with this flow:
  1. Overlay view starts as non-focusable (`FLAG_NOT_FOCUSABLE`), allowing normal interaction with the underlyring screens.
  2. The invoke keyboard button is pressed by the user.
  3. The overlay view layout params is switched to `FLAG_NOT_TOUCH_MODAL`, allowing it to be focused.
  4. The overlay view requests focus, then the soft keyboard is invoked with `inputMethodManager.toggleSoftInput`
  5. The overlay view is switched back to non-focusable, allowing normal interaction with the underlyring screens.
- The overlay button remains on screen and re-usable until manually dismissed, which allows the keyboard to be triggered at will multiple times.
- Applied all the other changes you have suggested.

Here is another video where I attempt to use this new version on multiple apps on my Android 13 device, you can see it works for any app below target SDK 33 (the bottom group), while failing for those above (the top group). I did also test it on my Android 9 device and it works fine in all cases. I also demonstrate the new interface, switching between apps and invoking the keyboard with touches visible. 

https://github.com/Julow/Unexpected-Keyboard/assets/39025047/1da21704-e30e-469e-a355-42376025e24c

Also replying @Spike-from-NH : 
1. Not at the moment
7. Adjusted
8. Yes
9. See (5)
10. Let's see if its still worth including first considering the new limitations found.

diejuse commented:
What about this solution??? https://www.youtube.com/watch?v=OCTiacmVu_Y&ab_channel=ArtTV%3AAGorgeousPlace
This keyboard is permanent and works even on Android 14. I just discovered it and made a comment on the video. I think it's a perfect solution but you need to have more options.

RetrogisusDEV commented:
I think that a good solution would be to use accessibility, to invoke the keyboard without the need for any text field, and it would be more shareable with previous APIs, than how accessibility is implemented? I don't know, but it would be the most viable solution for previous api (estimated up to api 19)

diejuse commented:
> I think that a good solution would be to use accessibility, to invoke the keyboard without the need for any text field, and it would be more shareable with previous APIs, than how accessibility is implemented? I don't know, but it would be the most viable solution for previous api (estimated up to api 19)

Yes that's how it is! I don't understand why there isn't any keyboard of this type! (apart from this). It would be perfect to have a floating icon above the rest of the apps and be able to display it whenever you want to send letters.
It would even be usable as a gamepad!

doak commented:
I am very much looking forward to this feature.

diejuse commented:
Anyone knows a always visible keyboard like this for Android? I need it.

LostRuins commented:
I will be closing this PR due to lack of interest. 

For future readers, this approach only works for apps with target sdk < 33.
For anyone who wants a fully working solution on Android 13+, use Game Keyboard 2.

diejuse commented:
> to invoke the keyboard without the need for any text field,

Couldn't you try adding the keyboard functionality you mentioned to your keyboard so we could use your wonderful keyboard at any time? 
That wat the keyboard could be to invoke without the need for any text field, and it would be more shareable with previous APIs, as said @RetrogisusDEV

diejuse commented:
> Sorry for the lack of interest. The target sdk constraint is a big problem as Google Play would not accept upgrades and would lower the rank of the app.
> 
> Though, I'd accept to release the app with this on F-Droid and Github, for a bit of time. It'd need to be made optional.
> 
> This do not work on my device (running Android 13): the keyboard seems to open and close immediately, no logs. Do you think there's a way to detect when it wouldn't work and not offer the feature in this case ?

Couldn't you try adding the keyboard functionality you mentioned to your keyboard so we could use your wonderful keyboard at any time?
That wat the keyboard could be to invoke without the need for any text field, and it would be more shareable with previous APIs, as said @RetrogisusDEV

RetrogisusDEV commented:
> > Sorry for the lack of interest. The target sdk constraint is a big problem as Google Play would not accept upgrades and would lower the rank of the app.
> > 
> > Though, I'd accept to release the app with this on F-Droid and Github, for a bit of time. It'd need to be made optional.
> > 
> > This do not work on my device (running Android 13): the keyboard seems to open and close immediately, no logs. Do you think there's a way to detect when it wouldn't work and not offer the feature in this case ?
> 
> @Julow Couldn't you try adding the keyboard functionality you mentioned to your keyboard so we could use your wonderful keyboard at any time?
> That wat the keyboard could be to invoke without the need for any text field, and it would be more shareable with previous APIs, as said @RetrogisusDEV

In fact, I've been trying and I'm already giving up xD
